libs/full/parcelport_mpi/include/hpx/parcelport_mpi/receiver_connection.hpp

```diff
@@ -152,19 +154,74 @@ namespace hpx::parcelset::policies::mpi {
 
         bool receive_chunks(std::size_t num_thread = -1)
         {
-            while (chunks_idx_ < buffer_.chunks_.size())
+            if (!request_done())
+            {
+                return false;
+            }
+
+            // handle zero-copy receive, this should be done on the first entry
+            // to receive_chunks only
+            if (parcels_.empty())
+            {
+                HPX_ASSERT(zero_copy_chunks_idx_ == 0);
+
+                auto const num_zero_copy_chunks = static_cast<std::size_t>(
+                    static_cast<std::uint32_t>(buffer_.num_chunks_.first));
+                if (num_zero_copy_chunks != 0)
+                {
+                    HPX_ASSERT(buffer_.chunks_.size() == num_zero_copy_chunks);
+
+                    // De-serialize the parcels such that all data but the
+                    // zero-copy chunks are in place. This de-serialization also
+                    // allocates all zero-chunk buffers and stores those in the
+                    // chunks array for the subsequent networking to place the
+                    // received data directly.
+                    for (std::size_t i = 0; i != num_zero_copy_chunks; ++i)
+                    {
+                        auto const chunk_size = static_cast<std::size_t>(
+                            buffer_.transmission_chunks_[i].second);
+                        buffer_.chunks_[i] =
+                            serialization::create_pointer_chunk(
+                                nullptr, chunk_size);
+                    }
+
+                    parcels_ =
+                        decode_parcels_zero_copy(pp_, buffer_, num_thread);
+
+                    // note that at this point, buffer_.chunks_ will have
+                    // entries for all chunks, including the non-zero-copy ones
+                }
+
+                // we should have received at least one parcel if there are
+                // zero-copy chunks to be received
+                HPX_ASSERT(parcels_.empty() || !buffer_.chunks_.empty());
+            }
+
+            while (chunks_idx_ != buffer_.chunks_.size())
             {
+                auto& c = buffer_.chunks_[chunks_idx_++];
+                if (c.type_ == serialization::chunk_type::chunk_type_index)
+                {
+                    continue;    // skip non-zero-copy chunks
+                }
+
                 if (!request_done())
```

Comment:
I think you should put this `if (!request_done())` in front of `chunks_idx_++` of line 202. Otherwise, you will skip this zero-copy chunk if the receive is not completed.

Reply:
Good catch! This is what was causing the assertion failures you mentioned above.